### ISSUE TITLE
how to test the model (this is not actually an issue)

### ISSUE THREAD

**Issue by fstylianou** [vinbigdata-medical/MIDL2021-VinDr-RibCXR#5]
I have run the code after making some minor modification. 
The code finished all 200 epochs. 
The final line of the log file is like this:
Validation dice: 0.781957, val loss:0.963734 best: 0.829610

My problem is how to proceed with a test on an image that the model has not seen?

After running the code I see a folder with _weights_ and inside a file named _multi_unet_b0_DiceLoss.yaml.pth_
In the _outputs_ folder I do not see anything. Is this normal?

If you have any Jupiter notebook I can use to visualise a testing Xray that will be great!

**Comment by levi3001** [comment]
you can refer to the visualization notebook to visualize (use for visualize the grouth truth) to visualize the output

**Comment by fstylianou** [comment]
I did that. But what about a test case?

After looking at the code I think I need to use something like the following:
`python main.py --config cvcore/config/multi_unet_b0_diceloss.yaml --load weights/multi_unet_b0_DiceLoss.yaml.pth --mode test ...`
but what exactly? and where to set the test X-ray image?

Update: I see that the test mode is not fully developed.

**Comment by levi3001** [comment]
yes, since there is a limit amount of data, we don't have test set

**Comment by fstylianou** [comment]
Sorry for asking again. I see that in the short paper MIDL2021 you have tested the code on two images that are not in the validation and training (i.e. the Shenzhen and JSRT). Do you happen to have the code that does that part? basically loads the weights and predicts the masks for the ribs.
I am trying to write a testing code and I am not confident on what I wrote. 
thanks

**Comment by levi3001** [comment]
I keep the code, I can send it to you via email

**Comment by fstylianou** [comment]
Thank you for replaying. I was able last night to make my testing code run. I will let you know if I need your version.
Thank you again.

**Comment by desalegnust** [comment]
Thank you for the wonderful baseline project. I was looking for the test code, can you share it to me?

**Comment by fstylianou** [comment]
I wrote the following basic python code to view the predictions of the model. Hope it helps...

```
import albumentations
from albumentations import Compose, Resize 
from albumentations.pytorch import ToTensorV2

import torch.nn as nn
import torch

import cv2
import numpy as np
import argparse
from PIL import Image

#from cvcore.config import get_cfg_defaults
from yacs.config import CfgNode as CN       #cvcore/config/default.py

#from cvcore.model import build_model
import segmentation_models_pytorch as smp   #cvcore/model/model_zoo.py


#from monai.metrics import DiceMetric
#from monai.losses import DiceLoss

cfg = CN()                  #cvcore/config/default.py
cfg.MODEL = CN()
cfg.MODEL.NAME = "unet(b0)" #cvcore/config/multi_unet_b0_diceloss.yaml
cfg.MODEL.NUM_CLASSES = 20  #cvcore/config/multi_unet_b0_diceloss.yaml
cfg.DATA = CN()
cfg.DATA.INP_CHANNEL = 1    #cvcore/config/multi_unet_b0_diceloss.yaml

#from cvcore/model/model_zoo.py
def build_model(cfg):
    if 'unet(b0)'== cfg.MODEL.NAME:
        model= smp.Unet('efficientnet-b0',classes=cfg.MODEL.NUM_CLASSES,encoder_weights='imagenet'\
        ,in_channels=cfg.DATA.INP_CHANNEL)
    return model


def get_args():
    # Define command line arguments
    parser = argparse.ArgumentParser(description="Test the trained model on a test image.")
    #parser.add_argument("--config", type=str, default="cvcore/config/multi_unet_b0_diceloss.yaml",
    #                    help="Path to the configuration file.")
    parser.add_argument("--weights", type=str, default="weights/multi_unet_b0_DiceLoss.yaml.pth",
                        help="Path to the trained weights file.")
    parser.add_argument("--image", type=str, 
                        #default="data/val/img/VinDr_RibCXR_val_013.png",
                        #default="data/val/img/VinDr_RibCXR_val_026.png", 
                        #default="/Users/ucy-compsci/U-Net lung segmentation/input/JSRT/pngs/JPCLN014.png", 
                        default="/Users/ucy-compsci/U-Net lung segmentation/input/shcxr-lung/images/CHNCXR_0015_0.png", 
                        help="Path to the test image file.")
    return parser.parse_args()


def load_image(image_path, transform):
    # Load the image and convert it to grayscale
    img = Image.open(image_path)
    img_size = img.size
    print(img_size)

    # Check the image mode
    #if mode is RGBA, then the image has four channels: Red, Green, Blue, and Alpha.
    #if mode is P, it means that the image is in a "palette" mode.
    print("Image mode:", img.mode)

    img = img.convert('L')

    # Check the image mode again
    print("Image mode:", img.mode) #mode is L (Luminance) indicates a grayscale image single-channel black and white.

    img = np.asarray(img, dtype=np.float32)
    img = img / 255
    img = np.expand_dims(img, axis=2)

    print("Image shape:", img.shape)

    # Apply data transformation (resize and convert to tensor)
    dic = transform(image=img)
    img_tensor = dic['image']

    ## Print the keys and values in the transformed dictionary
    #print("Type of dic:", type(dic))
    #print("Attributes and methods of dic:", dir(dic))
    #for key, value in dic.items():
    #    print(key, ":", type(value), value.shape)

    return img_tensor, img_size


def main(args):
    #print("Load config file...")
    #cfg = get_cfg_defaults()
    #cfg.merge_from_file(args.config)

    print("Build model...")
    model = build_model(cfg)

    ## Define loss
    #if cfg.LOSS.NAME == "ce":
    #    train_criterion = nn.CrossEntropyLoss()
    #elif cfg.LOSS.NAME=='Bce':
    #    train_criterion= nn.BCEWithLogitsLoss()
    #elif cfg.LOSS.NAME=='dice':
    #    train_criterion= DiceLoss(sigmoid=True)
    #
    #valid_criterion = train_criterion
    #if cfg.METRIC.NAME == 'dice':
    #    valid_metric = DiceMetric(include_background=True,reduction= 'mean')

    #model = model.to(device='cuda')   #fotos
    #model = model.to(device='cpu')    #fotos
    # Determine the target device for running the model
    if getattr(torch, 'has_mps', False):
        device = 'mps'
    elif torch.cuda.is_available():
        device = 'cuda'
    else:
        device = 'cpu'
    print(device)
    model = model.to(device)
    model = nn.DataParallel(model)

    print("Load weights...")
    checkpoint = torch.load(args.weights, "cpu")
    model.load_state_dict(checkpoint['state_dict'])
    print(f"Loaded checkpoint '{args.weights}' (epoch {checkpoint['epoch']})")

    print("Build data loader...")
    #data_loader = make_multi_ribs_dataloader(cfg, mode='test')
    data_transform = Compose([
        Resize(512, 512),
        ToTensorV2()
    ])
    img_tensor, img_size = load_image(args.image,transform=data_transform)

    print("Inference...")
    model.eval()

    list_label= ['R1','R2','R3','R4','R5','R6','R7','R8','R9','R10',\
                 'L1','L2','L3','L4','L5','L6','L7','L8','L9','L10']

    with torch.no_grad():
        img_tensor = img_tensor.to(device=device,dtype=torch.float)
        print("img_tensor shape:", img_tensor.shape)
        output = model(img_tensor.unsqueeze(0))
        print("output shape:", output.shape)
        output = torch.sigmoid(output) > 0.5
        output = output.cpu().numpy().astype(np.uint8)
        print(range(output.shape[1]))

        # Resize the output masks to the original image size and write to disk
        for i in range(output.shape[1]):
            mask = (output[0][i] * 255).astype(np.uint8)
            mask = cv2.resize(mask, (img_size[0], img_size[1]))
            cv2.imwrite("predictions/prediction_"+list_label[i]+".png", mask)
        
        print("Saved predictions")


if __name__ == '__main__':
    args = get_args()
    main(args)
```

**Comment by desalegnust** [comment]
Thank you so much for reaching out to me and extending your help.

I really appreciate your leading role in making a baseline project. I am
learning to understand how it works.

I will try the code

**Comment by fstylianou** [comment]
you should be thanking @levi3001 :)

**Comment by desalegnust** [comment]
Thankyou @levi3001  and the entire team behind VinDr-RibCXR
